src/app/calendar-app/runbox-calendar-event.ts

```diff
@@ -69,137 +88,437 @@ export class RunboxCalendarEvent implements CalendarEvent {
+        this._dtstart = startdate;
+        this._dtend = enddate;
+    }
 
-                if (!ICAL.TimezoneService.has(tz.tzid)) {
-                    ICAL.TimezoneService.register(tz.tzid, tz);
+    // An exception to the main (recurring) event
+    addExceptionEvent(
+        origdate: moment.Moment,
+        startdate: moment.Moment,
+        enddate: moment.Moment,
+        thisandfuture: boolean,
+        title: string,
+        description: string,
+        location: string): boolean {
+        if (this.isException) {
+            // This shouldnt be possible
+            console.log('Refusing to create an exception of an exception');
+            return false;
+        }
+        // clone existing one
+        const new_exception = new ICAL.Event(ICAL.Component.fromString(this.event.toString()));
+        new_exception.component.removeProperty('rrule');
+        const recurrence_id = this.momentToIcalTime(origdate, this.event.startDate.zone);
+        const new_start = this.momentToIcalTime(startdate, this.event.startDate.zone);
+        let new_end;
+        if (enddate) {
+            new_end = this.momentToIcalTime(enddate, this.event.startDate.zone);
+        }
+        if (this._dtstart.isDate) {
+            recurrence_id.isDate = true;
+            new_start.isDate = true;
+            new_end.isDate = true;
+        }
+        new_exception.recurrenceId = recurrence_id;
+        if (thisandfuture) {
+            const rId = new_exception.component.getFirstProperty('recurrence-id');
+            rId.setParameter('range', 'THISANDFUTURE');
+        }
+        new_exception.startDate = new_start;
+        if (new_end) {
+            new_exception.endDate = new_end;
+        }
+        if (title) {
+            new_exception.summary = title;
+        }
+        if (description) {
+            new_exception.description = description;
+        }
+        if (location) {
+            new_exception.location = location;
+        }
+        this.ical.addSubcomponent(new_exception.component);
+        // regenerate this.event to sort out the exceptions/recurrences?
+    }
+
+    // set values from edit dialog, inc logic for recurring
+    // parts + exception creation.
+    updateEvent(
+        dtstart:        moment.Moment,
+        dtend:          moment.Moment,
+        calendar:       string,
+        recur_save_type: string,
+        title:          string,
+        location:       string,
+        description:    string,
+        recurs:         boolean,
+        frequency:      string,
+        interval:       number,
+        recur_weekdays: string[],
+        recur_months:   string[],
+        recur_nth:      string[],
+    ) {
+        if (this.recurs && recur_save_type !== '0') {
+            // Adding an exception to an already recurring event
+            this.addExceptionEvent(
+                this.dtstart,
+                dtstart,
+                dtend,
+                recur_save_type === '2' ? true : false,
```

Comment:
Bits like these (and related in the editor dialog) make me think that this should be an enum, not a string :)

Reply:
Yup agreed, fixed